### PR TITLE
More robust handling of CORS HTTP Access Control

### PR DESCRIPTION
**Pull request by abeyad**
Uses a refactored version of Netty's CORS implementation to provide more
robust cross-origin resource request functionality.  The CORS specific
Elasticsearch parameters remain the same, just the underlying
implementation has changed.

It has also been refactored in a way that allows dropping in Netty's
CORS handler as a replacement once Elasticsearch is upgraded to Netty 4.

### CODE REVIEW

**Comment by abeyad** [comment]
@spinscale I made some changes as overriding the MessageEvent didn't work.  One change from how CORS handling previously worked is, before if an origin was not allowed, it would return a 200 but omit the "allow-origin" value in the response header.  Now, we return a 403 - this how Netty implements it too.

File: core/src/main/java/org/elasticsearch/http/netty/HttpUtils.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        return resp;
+    }
+
+    /**
+     * Convert a string representation of an HTTP method to its {@link HttpMethod} equivalent.
+     *
+     * @param method the method name to convert
+     * @return an {@link Optional} that contains the HttpMethod, or empty if no match was found.
+     */
+    public static Optional<HttpMethod> toHttpMethod(final String method) {
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/http/netty/HttpUtils.java]: Cant we just use `HttpMethod.valueOf()` instead of this? And thus also get rid of the `enLowerCase()` method above?

File: core/src/main/java/org/elasticsearch/http/netty/NettyHttpServerTransport.java [hunk 2]
```diff
@@ -312,6 +324,41 @@ protected void doStart() {
         this.boundAddress = new BoundTransportAddress(boundAddresses.toArray(new TransportAddress[boundAddresses.size()]), new InetSocketTransportAddress(publishAddress));
     }
 
+    private CorsConfig buildCorsConfig(Settings settings) {
+        if (!settings.getAsBoolean(SETTING_CORS_ENABLED, false)) {
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/http/netty/NettyHttpServerTransport.java]: `== false` here, as we do this in most other code as well

File: core/src/main/java/org/elasticsearch/http/netty/NettyHttpServerTransport.java [hunk 3]
```diff
@@ -312,6 +324,41 @@ protected void doStart() {
...
+        if (settings.getAsBoolean(SETTING_CORS_ALLOW_CREDENTIALS, false)) {
+            builder.allowCredentials();
+        }
+        String[] strMethods = settings.getAsArray(SETTING_CORS_ALLOW_METHODS, new String[0]);
+        HttpMethod[] methods = Arrays.asList(strMethods)
+                                     .stream()
+                                     .map(m -> HttpUtils.toHttpMethod(m))
+                                     .filter(Optional::isPresent)
+                                     .map(Optional::get)
+                                     .toArray(size -> new HttpMethod[size]);
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/http/netty/NettyHttpServerTransport.java]: looks more tedious than just filling an array, but I didnt do an awful lot of java 8 stuff yet :-)

File: core/src/test/java/org/elasticsearch/rest/CorsAnyOriginIT.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder()
+                   .put(Node.HTTP_ENABLED, true)
+                   .put(SETTING_CORS_ENABLED, true)
+                   .put(SETTING_CORS_ALLOW_ORIGIN, ANY_ORIGIN)
+                   .put(super.nodeSettings(nodeOrdinal)).build();
+    }
+
+    public void testThatCorsEnabledReturnsAnyOrigin() throws Exception {
+        HttpResponse response = httpClient().method("GET")
```
Comment (spinscale) [inline core/src/test/java/org/elasticsearch/rest/CorsAnyOriginIT.java]: Do you gain anything here compared to a unit test?

File: core/src/test/java/org/elasticsearch/rest/CorsOriginsIT.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                   .put(SETTING_CORS_ALLOW_METHODS, "get, options, post")
+                   .put(SETTING_CORS_ALLOW_CREDENTIALS, true)
+                   .put(SETTING_CORS_ENABLED, true)
+                   .put(Node.HTTP_ENABLED, true)
+                   .build();
+    }
+
+    @Override
+    public void testThatRegularExpressionWorksOnMatch() throws Exception {
+        String corsValue = "http://localhost:9200";
```
Comment (spinscale) [inline core/src/test/java/org/elasticsearch/rest/CorsOriginsIT.java]: same here, a unit test starting the netty transport should be sufficient and much faster?

**Comment by spinscale** [comment]
left a few comments, I think we can speed up tests and clean it up a little. Is this rebased against the latest master? Just want to make sure all the settings changed are in, and you're not surprised by those

File: core/src/main/java/org/elasticsearch/http/netty/HttpUtils.java (hunk 1 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/http/netty/HttpUtils.java]: This is a good point - I can get rid of the enLowerCase() anyway and toUpperCase() the input if I want case insensitivity.

File: core/src/test/java/org/elasticsearch/rest/CorsAnyOriginIT.java (hunk 4 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/rest/CorsAnyOriginIT.java]: @spinscale I did it more to maintain consistency with the CorsRegexIT test that was already in there, as well as the CorsNotSetIT.  But I can just as easily test the functionality by moving it to the netty transport tests.  I will do that for this and CorsOriginIT.

**Comment by abeyad** [comment]
@spinscale I implemented the suggested changes.  If you could take a look and let me know if its ready to merge.  Thanks!

File: core/src/main/java/org/elasticsearch/common/Strings.java [hunk 6]
```diff
@@ -36,6 +36,7 @@
 import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Locale;
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/common/Strings.java]: unneeded changes here

File: core/src/main/java/org/elasticsearch/http/netty/HttpRequestHandler.java [hunk 7]
```diff
@@ -68,9 +65,11 @@ public void messageReceived(ChannelHandlerContext ctx, MessageEvent e) throws Ex
         // when reading, or using a cumalation buffer
         NettyHttpRequest httpRequest = new NettyHttpRequest(request, e.getChannel());
         if (oue != null) {
-            serverTransport.dispatchRequest(httpRequest, new NettyHttpChannel(serverTransport, httpRequest, corsPattern, oue, detailedErrorsEnabled));
+            serverTransport.dispatchRequest(httpRequest,
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/http/netty/HttpRequestHandler.java]: I think you can leave it in one line, no need to break after 80 chars or sth (unless I missed that :-)

File: core/src/main/java/org/elasticsearch/rest/support/RestUtils.java [hunk 8]
```diff
@@ -238,4 +240,21 @@ public static Pattern checkCorsSettingForRegex(String corsSetting) {
...
+    /**
+     * Return the CORS setting as an array of origins.
+     *
+     * @param corsSetting the CORS allow origin setting as configured by the user;
+     *                    should never pass null, but we check for it anyway.
+     * @return an array of origins if set, otherwise {@code null}.
+     */
+    public static String[] corsSettingAsArray(String corsSetting) {
+        if (Strings.isNullOrEmpty(corsSetting)) {
+            return null;
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/rest/support/RestUtils.java]: would an empty array be easier to handle to prevent a non-meaning null return or would that complicate things?

File: core/src/test/java/org/elasticsearch/common/StringsTests.java [hunk 9]
```diff
@@ -60,4 +60,5 @@ public void testCleanTruncate() {
         assertEquals("o", Strings.cleanTruncate("o\uD83D\uDEAB", 1));
         assertEquals("", Strings.cleanTruncate("foo", 0));
     }
+
```
Comment (spinscale) [inline core/src/test/java/org/elasticsearch/common/StringsTests.java]: unneeded (bloats the diff)

File: core/src/test/java/org/elasticsearch/rest/CorsNotSetIT.java [hunk 10]
```diff
@@ -31,7 +32,8 @@
 /**
  *
  */
-public class CorsRegexDefaultIT extends ESIntegTestCase {
+@ClusterScope(scope = ESIntegTestCase.Scope.SUITE, numDataNodes = 1)
```
Comment (spinscale) [inline core/src/test/java/org/elasticsearch/rest/CorsNotSetIT.java]: was this change intentional or did you just want to rename?

**Comment by spinscale** [comment]
left minor comments, apart from that LGTM.

I think it might make sense to take the next step here and move netty into its own module, then we can easily get rid of this again in our own codebase once we upgrade to netty4, by using the netty cors handler.

**Comment by abeyad** [comment]
From what I could tell, the Netty CORS handler does not support regular expressions, which is something ES supports.  So I had to modify the Netty CORS handler as such and if we want to continue to support regular expressions for the origin value (other than "*" for any origin), then we can't just use a drop in replacement (unless of course Netty is adding regular expression capability to their CORS handler).

File: core/src/test/java/org/elasticsearch/rest/CorsNotSetIT.java (hunk 10 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/rest/CorsNotSetIT.java]: I just figured to be explicit, as thats the default but makes it clearer when its spelled out.  Also, the numDataNodes is selected randomly by default, but I didn't see how that would benefit this test, and the other similar tests don't have this.

File: core/src/main/java/org/elasticsearch/rest/support/RestUtils.java (hunk 8 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/rest/support/RestUtils.java]: Yes, empty array would be easier and more consistent with how its handled in the NettyTransport.  I'll make the change.

**Comment by abeyad** [comment]
@spinscale I pushed up changes and also rebased with master, let me know what you think.

**Comment by spinscale** [comment]
tests passed on local run, LGTM
